pkg/operator/controllers/machine/machine_controller.go

```diff
@@ -0,0 +1,94 @@
+package machine
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"strings"
+
+	machinev1beta1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	maoclient "github.com/openshift/machine-api-operator/pkg/generated/clientset/versioned"
+	"github.com/operator-framework/operator-sdk/pkg/status"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/Azure/ARO-RP/pkg/operator"
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+	_ "github.com/Azure/ARO-RP/pkg/util/scheme"
+)
+
+type MachineReconciler struct {
+	maocli                 maoclient.Interface
+	arocli                 aroclient.Interface
+	log                    *logrus.Entry
+	isLocalDevelopmentMode bool
+	role                   string
+}
+
+func NewMachineReconciler(log *logrus.Entry, maocli maoclient.Interface, arocli aroclient.Interface, isLocalDevelopmentMode bool, role string) *MachineReconciler {
+	return &MachineReconciler{
+		maocli:                 maocli,
+		arocli:                 arocli,
+		log:                    log,
+		isLocalDevelopmentMode: isLocalDevelopmentMode,
+		role:                   role,
+	}
+}
+
+func (r *MachineReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	// Update cluster object's status.
+	cond := &status.Condition{
+		Type:    arov1alpha1.MachineValid,
+		Status:  corev1.ConditionTrue,
+		Message: "All machines valid",
+		Reason:  "CheckDone",
+	}
+
+	errs := r.checkMachines(ctx)
+	if len(errs) > 0 {
+		cond.Status = corev1.ConditionFalse
+		cond.Reason = "CheckFailed"
+
+		var sb strings.Builder
+		for _, err := range errs {
+			sb.WriteString(err.Error())
+			sb.WriteByte('\n')
+		}
+		cond.Message = sb.String()
+	}
+
+	err := controllers.SetCondition(ctx, r.arocli, cond, r.role)
+	if err != nil {
+		r.log.Error(err)
+	}
+
+	return reconcile.Result{}, err
+}
+
+// Setup our mananger
+func (r *MachineReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	aroClusterPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetName() == arov1alpha1.SingletonClusterName
+	})
+
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}, builder.WithPredicates(aroClusterPredicate))
```

Comment:
I think we do not need to watch for this object: we are only watching for machine changes, right?

I imagine it should be something simply as


```go

        func (r *MachineReconciler) SetupWithManager(mgr ctrl.Manager) error {
        	return ctrl.NewControllerManagedBy(mgr).
        		For(machinev1beta1.Machine{}).
        		Named(controllers.MachineControllerName).
        		Complete(r)
        }
```

Reply:
But probably with `if r.role == operator.RoleMaster`

Reply:
I've realized that since we are breaking away the former `machinechecker.go` from the `Checker` controller, we should no longer need `if r.role == operator.RoleMaster` since that logic is now in `operator.go.` The SetupWithManager() has been simplified according to your suggestions.

Reply:
@cadenmarchese hm... Are you sure about `if r.role == operator.RoleMaster`? Without this we will be running two reconcilers: one on one of the master nodes and another on one of the worker nodes. There (probably) will be a rance condition between two of them on an attempt to update status of the object. It will not manifest loudly due to retries, but if we can avoid it - we should.

Or do I miss something?

Reply:
Agreed, and now that we have that logic in operator.go, we shouldn't need it here - machine controller will not run if operator role is worker: https://github.com/Azure/ARO-RP/blob/0b81a3f4c81b0b3aa0b5155e53676e12ba761e98/cmd/aro/operator.go#L99 https://github.com/Azure/ARO-RP/blob/0b81a3f4c81b0b3aa0b5155e53676e12ba761e98/cmd/aro/operator.go#L166-L170